### DISCUSSION TITLE
Running the cusvaer benchmark on multiple hosts, each has >=1 GPU

### DISCUSSION THREAD

**Discussion by rht** [NVIDIA/cuQuantum#114] [Q&A] [answered]
It has been a while. I am looking to run one of the example commands in https://github.com/NVIDIA/cuQuantum/tree/main/benchmarks,
```
mpiexec -n 4 cuquantum-benchmarks circuit --frontend qiskit --backend cusvaer --benchmark quantum_volume --nqubits 32 --ngpus 1 --cusvaer-global-index-bits 1,1 --cusvaer-p2p-device-bits 1
```
but on multiple hosts (i.e. `--hostfile hosts`). The [answer](https://github.com/NVIDIA/cuQuantum/discussions/37#discussioncomment-6909799) to https://github.com/NVIDIA/cuQuantum/discussions/37 is not sufficient.

On a fresh AWS instance, fresh cuQuantum Appliance 23.10,
```
docker run -it --detach --name client_${client_hostname}_0 --gpus "'device=0'" \
    --network=host \
    --uts=host \
    --ipc=host \
    nvcr.io/nvidia/cuquantum-appliance:23.10
```

I did steps that deviate from the answer from discussion #37:
- `PubkeyAuthentication yes` instead of `PermitRootLogin yes`
- Uncommented `HostKey /etc/ssh/ssh_host_ed25519_key`, which is mandatory but not said in the answer. (Maybe the default config in 23.10 has changed since 23.06?)

I got the following error:
```
+ /usr/bin/mpirun --hostfile hosts -v -n 2 --allow-run-as-root hostname
--------------------------------------------------------------------------
A compressed message was received by the Open MPI run time system
(PMIx) that could not be decompressed.  This means that Open MPI has
compression support enabled on one node and not enabled on another.
This is an unsupported configuration.

Compression support is enabled when both of the following conditions
are met:

1. The Open MPI run time system (PMIx) is built with compression
   support.
2. The necessary compression libraries (e.g., libz) can be found at
   run time.

You should check that both of these conditions are true on both the
node where mpirun is invoked and all the nodes where MPI processes
will be launched.  The node listed below does not have both conditions
met:

  node without compression support:  ip-172-31-35-127

NOTE: There may also be other nodes without compression support.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
ORTE was unable to reliably start one or more daemons.
This usually is caused by:

* not finding the required libraries and/or binaries on
  one or more nodes. Please check your PATH and LD_LIBRARY_PATH
  settings, or configure OMPI with --enable-orterun-prefix-by-default

* lack of authority to execute on one or more specified nodes.
  Please verify your allocation and authorities.

* the inability to write startup files into /tmp (--tmpdir/orte_tmpdir_base).
  Please check with your sys admin to determine the correct location to use.

*  compilation of the orted with dynamic libraries when static are required
  (e.g., on Cray). Please check your configure cmd line and consider using
  one of the contrib/platform definitions for your system type.

* an inability to create a connection back to mpirun due to a
  lack of common network interfaces and/or no route found between
  them. Please check network connectivity (including firewalls
  and network routing requirements).
```
In case you wonder why the command is `/usr/bin/mpirun --hostfile hosts -v -n 2 --allow-run-as-root hostname` instead of the benchmark script, it is because it fails even at the hostname check, and running the benchmark script has the same error message.

Have you seen the error message and resolved it?

**Comment by rht** [comment]
I have ensured that the package `zlib1g` and `zlib1g-dev` are installed on the 2 nodes I have set up.

**Comment by ymagchi** [comment]
Hi @rht,
Would it be possible to share the location of mpirun when `which mpirun` is executed?

**Reply by rht** [reply]
This is the output of `which mpirun`: `/usr/local/openmpi/bin/mpirun`. The `/usr/bin/mpirun` is from `sudo apt install openmpi-bin openmpi-doc libopenmpi-dev`. I have removed the Ubuntu-packaged openmpi, and instead ran
```
exe=/usr/local/openmpi/bin/mpirun
$exe --hostfile hosts -v -n 2 --allow-run-as-root hostname
```
This is the error message I got
```
bash: /usr/local/openmpi/bin/orted: No such file or directory
--------------------------------------------------------------------------
ORTE was unable to reliably start one or more daemons.
This usually is caused by:

* not finding the required libraries and/or binaries on
  one or more nodes. Please check your PATH and LD_LIBRARY_PATH
  settings, or configure OMPI with --enable-orterun-prefix-by-default

* lack of authority to execute on one or more specified nodes.
  Please verify your allocation and authorities.

* the inability to write startup files into /tmp (--tmpdir/orte_tmpdir_base).
  Please check with your sys admin to determine the correct location to use.

*  compilation of the orted with dynamic libraries when static are required
  (e.g., on Cray). Please check your configure cmd line and consider using
  one of the contrib/platform definitions for your system type.

* an inability to create a connection back to mpirun due to a
  lack of common network interfaces and/or no route found between
  them. Please check network connectivity (including firewalls
  and network routing requirements).
--------------------------------------------------------------------------
```
I looked at `/usr/local/openmpi/bin` of both Docker nodes, and they do indeed have `orted`.

**Reply by rht** [reply]
I did also install from `wget https://download.open-mpi.org/release/open-mpi/v5.0/openmpi-5.0.1.tar.bz2`, possibly causing havoc, even after I had done `sudo make uninstall` on it. Let me try to redo everything from a fresh cuQuantum Appliance 23.10.

**Reply by rht** [reply]
Confirmed that after redoing everything from scratch, without openmpi from Ubuntu packages, and openmpi from open-mpi.org, I got the error message of https://github.com/NVIDIA/cuQuantum/discussions/114#discussioncomment-8330938, i.e. `bash: /usr/local/openmpi/bin/orted: No such file or directory`.

**Reply by ymagchi** [reply]
Thank you for the information.
Assuming that `/usr/local/openmpi/bin/orted` exists on all the nodes, could you please verify that your PATH and LD_LIBRARY_PATH are set correctly on each remote host?

**Reply by rht** [reply]
`/usr/local/openmpi/lib` seems missing in `$LD_LIBRARY_PATH`

```
$ echo $PATH
/home/cuquantum/conda/envs/cuquantum-23.10/bin:/opt/conda/condabin:/home/cuquantum/conda/envs/cuquantum-23.10/bin:/opt/conda/bin:/usr/local/mpi/bin:/usr/local/nvidia/bin:/usr/local/openmpi/bin:/usr/local/pmix/bin:/usr/local/nvidia/bin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin

$ echo $LD_LIBRARY_PATH
/usr/local/pmix/lib:/usr/local/nvidia/lib:/usr/local/nvidia/lib64
```

After checking the env vars above, I added `export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/openmpi/lib` to both instances, relogin-ed, and reran
```
exe=/usr/local/openmpi/bin/mpirun
$exe --hostfile hosts -v -n 2 --allow-run-as-root hostname
```
I got the same error message (`bash: /usr/local/openmpi/bin/orted: No such file or directory`).

---

However, when I ran
```
exe=mpirun
$exe --hostfile hosts -v -n 2 --allow-run-as-root hostname
```
I got the original error message I posted in this discussion: https://github.com/NVIDIA/cuQuantum/discussions/114#discussion-6154233

---

I figured out why it has different version of `mpirun`. It seems that it tries to `ssh` to the second host's host OS instead of the Docker instance. And I made a mistake in specifying this in my ~/.ssh/config
```
Host hostname-of-first-host
        HostName 54.x.x.x
        user ubuntu
Host client_host1_0
        HostName 54.x.x.x
        User cuquantum
        Port 2383
Host client_host2_0
        HostName 52.x.x.x
        User cuquantum
        Port 2383
```
My hostfile contains `hostname-of-first-host` instead of client_host1_0

**Reply by rht** [reply, marked as answer]
Update: it works now! Thank you for helping to narrow down the causes (and that it must be an error from my configuration). New mistakes I had to fix:
- Should be client-host1-0, client-host2-0 (dash instead of underscore)
- Should be explicit `/usr/local/openmpi/bin/mpirun` instead of `mpirun`
- The hostfile should contain `localhost` for self (instead of client-host1-0) and then `client-host2-0`

This is the `which mpirun` of the host OS: `/opt/amazon/openmpi/bin/mpirun`

**Reply by rht** [reply]
@ymagchi thank you once again. If I want to submit a PR documenting all the steps to do the benchmark on multiple hosts, where should I add the how-to snippet to?

**Reply by ymagchi** [reply]
I'm glad you could figure things out.

At this time we do not accept code contributions (ref: [CONTRIBUTING.html](https://docs.nvidia.com/cuda/cuquantum/latest/python/CONTRIBUTING.html)), but we are planning to update our documentation about the detailed usage in the cloud. cc @mtjrider for vis.
